train_discreteaction_pendulum.py

```diff
@@ -0,0 +1,242 @@
+    """This function plots the learning curve as the network is training or shows the complete learning curve"""
+    plt.figure(1)
+    if show_result:
+        plt.title('Result')
+    else:
+        plt.clf()
+        plt.title('Training')
+    plt.xlabel('Episode')
+    plt.ylabel('Episode Return')
+    plt.plot(episode_returns)
+    plt.grid()
+    plt.ylim([-10, 110])
+    # Take past x episode returns plot the average
+    if len(episode_returns) >= mean_num and not show_result:
+        episode_means.append(np.mean(episode_returns[-mean_num:-1]))
+    elif not show_result:
+        episode_means.append(np.sum(episode_returns)/mean_num)
+    plt.plot(range(len(episode_means)), episode_means)
+    plt.pause(0.0001)  # pause a bit so that plots are updated
+    if is_ipython:
+        if not show_result:
+            display.display(plt.gcf())
+            display.clear_output(wait=True)
+        else:
+            display.display(plt.gcf())
+    return episode_means
+
+
+def optimize_model():
+    #Optimize model with random set of transition memories
+    if len(memory) < BATCHSIZE:
+        #skip if not enough memories for batch
+        return
+    transitions = memory.sample(BATCHSIZE)  #The sampled memories
+    batch = Transition(*zip(*transitions))  #This takes the list of Transition tuples and makes a single Transition tuple of all of the transitions. i.e. batch.action gives all the actions as a tuple while transitions[i].action gives the action of the ith  transition
+
+    non_final_mask = torch.tensor(tuple(map(lambda s: s is not None,
+                                            batch.next_state)), device=device, dtype=torch.bool) #Get a mask transitions that do not lead to a terminal state
+    
+    non_final_next_states = torch.cat([s for s in batch.next_state
+                                       if s is not None])
+
+    state_batch = torch.cat(batch.state)                #Tensor of states
+    action_batch = torch.cat(batch.action)              #Tensor of actions
+    reward_batch = torch.cat(batch.reward)              #Tensor or rewards
+
+    state_action_values = Q_net(state_batch).gather(1, action_batch) #the Q values for the state of the current batch of memories
+    next_state_values = torch.zeros(BATCHSIZE) 
+    with torch.no_grad():
+        next_state_values[non_final_mask] = target_net(
+            non_final_next_states).max(1)[0]                        #The next state values are the max the target network for the next states
+
+    # Compute the expected Q values
+    expected_state_action_values = (next_state_values * GAM) + reward_batch     #
+
+    # Loss
+    optimizer.zero_grad()
+    criterion = nn.SmoothL1Loss()
+    loss = criterion(state_action_values,
+                     expected_state_action_values.unsqueeze(1))
+
+    # Optimize the model
+    loss.backward()
+    # In-place gradient clipping
+    torch.nn.utils.clip_grad_value_(Q_net.parameters(), 100)
```

Comment:
I was unsure about this if the gradients need to be clipped during or after the back propagation and if that made a difference.

Reply:
I think you're right they should be before the back prop. Though I'm guessing the clipping after is just making them be clipped for the next step and still prevents the exploding grads